sled-storage/src/manager.rs

```diff
@@ -617,429 +836,520 @@ impl StorageManager {
+    #[tokio::test]
+    async fn ledgerless_to_ledgered_migration() {
+        illumos_utils::USE_MOCKS.store(false, Ordering::SeqCst);
+        let logctx = test_setup_log("ledgerless_to_ledgered_migration");
+        let mut harness = StorageManagerTestHarness::new(&logctx.log).await;
+
+        // Test setup: Create a U.2 and M.2
+        let raw_disks =
+            harness.add_vdevs(&["u2_under_test.vdev", "m2_helping.vdev"]).await;
+
+        // First, we format the U.2 to have a zpool. This should work, even
+        // without looping in the StorageManager.
+        let u2_raw = &raw_disks[0];
+        let pool_id = Uuid::new_v4();
+        let _disk = crate::disk::Disk::new(
+            &logctx.log,
+            &harness.mount_config(),
+            u2_raw.clone(),
+            Some(pool_id),
+            Some(harness.key_requester()),
+        )
+        .await
+        .expect("Failed to format U.2");
+
+        // Because we did that formatting "behind the back" of the
+        // StorageManager, we should see no evidence of the U.2 being managed.
+        //
+        // This currently matches the format of "existing systems, which were
+        // initialized before the storage ledger was created".
+
+        // We should still see no ledger.
+        let result = harness.handle().omicron_physical_disks_list().await;
+        assert!(matches!(result, Err(Error::LedgerNotFound),), "{:?}", result);
+
+        // We should also not see any managed U.2s.
+        let disks = harness.handle().get_latest_disks().await;
+        assert!(disks.all_u2_zpools().is_empty());
+
+        // However, when the system activates, we should see a Zpool,
+        // and "auto-manage" it.
+
+        harness.handle().key_manager_ready().await;
+
+        // It might take a moment for synchronization to be handled by the
+        // background task, but we'll eventually see the U.2 zpool.
+        //
+        // This is the equivalent of us "loading a zpool, even though
+        // it was not backed by a ledger".
+        let handle = harness.handle_mut();
+        while handle.wait_for_changes().await.all_u2_zpools().is_empty() {
+            info!(&logctx.log, "Waiting for U.2 to automatically show up");
+        }
 
-        // Create and add a disk
-        let zpool_name = ZpoolName::new_external(Uuid::new_v4());
-        let dir = tempdir().unwrap();
-        let disk: RawDisk =
-            SyntheticDisk::create_zpool(dir.path(), &zpool_name, 0).into();
-        handle.upsert_disk(disk.clone()).await;
+        // This is the equivalent of the "/omicron-physical-disks GET" API,
+        // which Nexus might use to contact this sled.
+        //
+        // This means that we'll bootstrap the sled successfully, but report a
+        // 404 if nexus asks us for the latest configuration.
+        let result = harness.handle().omicron_physical_disks_list().await;
+        assert!(matches!(result, Err(Error::LedgerNotFound),), "{:?}", result);
+
```

Comment:
I think this is the point where you could add a newly detected unmanaged disk by formatting it behind the back of the `StorageManager` similar to above. You could then wait for it to show up in `all_u2_zpools`. My guess is that without the change I suggested above to fix the race condition, this test will fail.

Reply:
I did something *kinda* like this, but it's a little tricky to ensure the synthetic disk gets formatted with a zpool. I'm accomplishing this by plumbing the ability to "attach/detach disks from the test harness".